### ISSUE TITLE
Parsing messes up when given a utf-8 input with a BOM

### ISSUE THREAD

**Issue by thislooksfun** [inikulin/parse5#936]
I have attached two files. One of them starts with a utf-8 BOM (`0xEF, 0xBB, 0xBF`), and the other does not. They are otherwise identical, and look like this:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    Hello world!
  </body>
</html>
```

They look the same in my editor, and they act the same when read and printed back out as files, but when I pass them to `serialize(parse(fileContents))` the file without the BOM prints

```html
<!DOCTYPE html><html lang="en"><head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
  </head>
  <body>
    Hello world!
  

</body></html>
```

This all makes sense so far. However, when I do the exact same process to the file with the BOM I get the output

```html
<html lang="en"><head></head><body>﻿

  
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
  
  
    Hello world!
  

</body></html>
```

As you can see, the contents of the `<head>` have magically moved down into the `<body>`.

[bom.html.txt](https://github.com/inikulin/parse5/files/11503789/bom.html.txt)
[bomless.html.txt](https://github.com/inikulin/parse5/files/11503791/bomless.html.txt)

**Comment by fb55** [comment]
Duplicate of https://github.com/inikulin/parse5/issues/111. Tl;dr input stream decoding is out of scope for parse5, but you could use eg. https://github.com/fb55/encoding-sniffer to decode the input before using parse5.

**Comment by thislooksfun** [comment]
This is not a duplicate of #111, that issue is for automatically decoding the input stream. All I'm asking is for string inputs to either parse properly (they currently don't) or throw an error. Undefined behavior, especially regarding invisible differences like a BOM is a bad developer UX. I only realized that this was occurring because I was closely comparing my test inputs and the re-serialized output. It would be really easy for a developer to pass a BOM-marked utf-8 string without realizing it (e.g. if some web server prefixes its responses with a BOM) and not notice that their output is wrong. Or worse, notice that it's wrong and not be able to find the issue.

TL;DR: Please either treat BOM-marked strings as valid and parse them correctly, treat them as invalid and throw an error.

**Comment by fb55** [comment]
This library implements the HTML parsing spec, which does not expect a BOM.

If you will only ever encounter an UTF8 BOM, then a simple `String.prototype.replace` call suffices (or a `.slice` if you know the BOM will always be present). A BOM generally point towards a need to support different encodings, which is out of scope for this module.

**Comment by wooorm** [comment]
slice/replace messes up positional info, which is also a very useful part of that this project provides which is outside of the scope of the HTML spec

**Comment by wooorm** [comment]
I think it’s an understandable request that, as this project is often used inside Node.js, and that users there often read HTML files from the file system or from over the network, and that the HTML spec mentions that users can start HTML files with BOMs (see for example § 13.1, 13.2.3), that BOMs are handled gracefully by this project or that, at least, it is explained clearly how to use this project from Node.js

**Comment by fb55** [comment]
I was thinking 13.2, but forgot that encoding handling is part of that section. Still, this is a quite complicated subject and should IMHO be handled by a separate module as a pre-processing step (this is what JSDOM and Cheerio do). I maintain https://github.com/fb55/encoding-sniffer as this pre-processing step.

> users there often read HTML files from the file system or from over the network

Agreed. But supporting a small subset of possible encodings (only UTF8 with a BOM) feels insufficient; having a proper split of responsibilities should make it easier for users to know what to expect.

> slice/replace messes up positional info, which is also a very useful part that this project provides which is outside of the scope of the HTML spec

This will always be an issue with character encodings — there is currently no easy way to map the original bytes to the code point positions in the input stream.

The expected byte positions also depend on the use-case. Code editors will ignore BOMs when displaying files, so document positions surfaced to users will be wrong if a BOM is taken into account. (And this applies to most character encodings.)

I don't feel like I have a good answer for how to deal with this — I'm open for suggestions.

> it’s an understandable request that, [...] at least, it is explained clearly how to use this project from Node.js

Strong yes. Let's figure out how to deal with this issue, then update the docs.

**Comment by wooorm** [comment]
I strongly prefer supporting **at least** UTF-8 BOM. Or even all 3 BOMs mentioned to the HTML spec.

- Users typically pass strings to this project, which will have to be decoded properly by them anyway. Or by your project. But those BOMs should still be in the string.
- The HTML spec tells users to always use UTF-8.

> feels insufficient

Why?

>  there is currently no easy way to map the code point positions in the input stream to the original bytes.

As far as I am aware, this project does not expose byte information in positional info, it instead uses character offset?
And it, except for a BOM, does *exactly* map positions to the original passed string?

**Comment by fb55** [comment]
> But those BOMs should still be in the string no?

They won't be — eg. iconv-lite strips BOMs.

> The HTML spec tells users to always use UTF-8.

The HTML spec uses unicode code points internally, but defaults to windows-1252 as the encoding for the vast majority of locales (see table at the bottom of https://html.spec.whatwg.org/multipage/parsing.html#determining-the-character-encoding).

> > feels insufficient
>
> Why?

For starters, because UTF8 isn't the default encoding for HTML.

> As far as I am aware, this project does not expose byte information in positional info, it instead uses character offsets the way they work in JS strings?

This project does expose line/col positions as well as offsets.

**Comment by wooorm** [comment]
Even if BOMs aren’t kept at the start, it looks like `DOMParser` treats BOMs as regular characters, and puts them in body:

```js
const parser = new DOMParser()
const doc = parser.parseFromString("\ufeff<title>hi</title><h1>hi!</h1>", "text/html");
doc.body.textContent.charCodeAt(0).toString(16); // 'feff'
```
